lib/internal/tls.js

```diff
@@ -0,0 +1,34 @@
+'use strict';
+
+const net = require('net');
+
+function canonicalIp(address) {
+  // Convert the ip address into the same format
+  // stored in certificates
+  if (net.isIPv6(address)) {
+    const b = ['0', '0', '0', '0', '0', '0', '0', '0'];
+
+    const s = address.split('::');
+    if (s.length === 2) {
+      const s1 = s[0].split(':');
+      for (var n = 0; n < s1.length; n++) {
+        if (s1[n]) {
+          b[n] = s1[n].replace(/^0+(\d+)$/, '$1');
+        }
+      }
+      const s2 = s[1].split(':');
+      for (n = 0; n < s2.length; n++) {
+        if (s2[n]) {
+          b[8 - s2.length + n] = s2[n].replace(/^0+(\d+)$/, '$1');
+        }
+      }
+    }
+
+    return b.join(':');
```

Comment:
If you want to improve the performance you could do the following to improve it further.
 - use string concatenation and not use the `b` array at all. And at the end add `"0:".repeat(restLen)`.
 - replace the RegExp with a loop that checks something like `var n = 0; while (str.charCodeAt(n++) === 48);` so you know what part you want to slice out.
 - use `address.indexOf('::')` and `address.split(':')` directly. You know until where the first part and until where the second part will go because of the indexOf and you only have to split a single time instead of three times (you could optimize this the same as the RegExp by using a loop and not splitting at all. It should be even faster).
- The `if (s1[n])` is not necessary as far as I see it. The current test cases have no entry that does not match but in case e.g. `:fe80::0000:` is a valid entry, just check the very first/last entry if it is entry and count down the length.

But this is out of my perspective not that important as a first draft and should not hold off the PR from being merged.

Reply:
This code is run once per https-request, and only if the host is specified with an IPv6 address.  The code is already quite complex as it is and I would not like to complicate it further right now.

The `if (s1[n])` is needed for '::1'